imagecraft/models/volume.py

```diff
@@ -55,13 +56,47 @@
     "and hyphens, and may not begin or end with a hyphen."
 )
 
+SIZE_INVALID_MSG = "size must be expressed in bytes, optionally with M or G unit."
+
 StructureName = typing.Annotated[
     str,
     StringConstraints(
         max_length=GPT_NAME_MAX_LENGTH, pattern=STRUCTURE_NAME_COMPILED_REGEX
     ),
 ]
 
+
+def _validate_structure_size(value: str) -> str:
+    """Validate and convert the structure size.
+
+    The Volume specification does not respect the IEC 80000-13 Standard:
+    M means MiB (2^20) in the Volume spec. (means 10^2 in the standard).
+    So this validation converts the unit before the conversion to the ByteSize
+    type (respecting the IEC 80000-13 Standard).
+    """
+    value = str(value)
+    match = STRUCTURE_SIZE_COMPILED_REGEX.match(value)
+
+    if not match:
+        raise ValueError(SIZE_INVALID_MSG)
+
+    # convert M/G to MiB/GiB
+    unit: str = ""
+    unit_matched = match.group("unit")
+    if unit_matched == "M":
+        unit = "MiB"
+    elif unit_matched == "G":
+        unit = "GiB"
```

Comment:
Personally I would just declare the field as numeric and resolve the value directly instead of translating this to something that pydantic understands to then let it convert values. Doing it ourselves is simpler, faster, and would yield the same result (we may already have the unit constants if we do size computation elsewhere in the code).